lib/collection/src/collection_manager/optimizers/config_mismatch_optimizer.rs

```diff
@@ -0,0 +1,492 @@
+
+                            false
+                        });
+
+                has_mismatch.then_some((*idx, vector_size))
+            })
+            .collect();
+
+        // Select segment with largest vector size
+        candidates
+            .into_iter()
+            .max_by_key(|(_, vector_size)| *vector_size)
+            .map(|(segment_id, _)| segment_id)
+            .into_iter()
+            .collect()
+    }
+}
+
+impl SegmentOptimizer for ConfigMismatchOptimizer {
+    fn collection_path(&self) -> &Path {
+        self.segments_path.as_path()
+    }
+
+    fn temp_path(&self) -> &Path {
+        self.collection_temp_dir.as_path()
+    }
+
+    fn collection_params(&self) -> CollectionParams {
+        self.collection_params.clone()
+    }
+
+    fn hnsw_config(&self) -> &HnswConfig {
+        &self.hnsw_config
+    }
+
+    fn quantization_config(&self) -> Option<QuantizationConfig> {
+        self.quantization_config.clone()
+    }
+
+    fn threshold_config(&self) -> &OptimizerThresholds {
+        &self.thresholds_config
+    }
+
+    fn check_condition(
+        &self,
+        segments: LockedSegmentHolder,
+        excluded_ids: &HashSet<SegmentId>,
+    ) -> Vec<SegmentId> {
+        self.worst_segment(segments, excluded_ids)
+    }
+
+    fn get_telemetry_data(&self) -> OperationDurationStatistics {
+        self.get_telemetry_counter().lock().get_statistics()
+    }
+
+    fn get_telemetry_counter(&self) -> Arc<Mutex<OperationDurationsAggregator>> {
+        self.telemetry_durations_aggregator.clone()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::BTreeMap;
+    use std::sync::Arc;
+
+    use parking_lot::RwLock;
+    use segment::entry::entry_point::SegmentEntry;
+    use segment::types::Distance;
+    use tempfile::Builder;
+
+    use super::*;
+    use crate::collection_manager::fixtures::{random_multi_vec_segment, random_segment};
+    use crate::collection_manager::holders::segment_holder::{LockedSegment, SegmentHolder};
+    use crate::collection_manager::optimizers::indexing_optimizer::IndexingOptimizer;
+    use crate::operations::config_diff::HnswConfigDiff;
+    use crate::operations::types::{VectorParams, VectorsConfig};
+
+    /// This test the config mismatch optimizer for a changed HNSW config
+    ///
+    /// It tests whether:
+    /// - the condition check for HNSW mismatches works
+    /// - optimized segments (and vector storages) use the updated configuration
+    ///
+    /// In short, this is what happens in this test:
+    /// - create randomized segment as base
+    /// - use indexing optimizer to build index for our segment
+    /// - test config mismatch condition: should not trigger yet
+    /// - change collection HNSW config
+    /// - test config mismatch condition: should trigger due to HNSW change
+    /// - optimize segment with config mismatch optimizer
+    /// - assert segment uses changed configuration
+    #[test]
+    fn test_hnsw_config_mismatch() {
```

Comment:
Nice summary! I believe we could improve these long tests by using additional tests to validate the intermediate steps, as opposed to chaining action - assertions together 😄

Reply:
Yeah, good idea. It is a bit _hard_ to directly construct these intermediate states, which is why I ended up with this create-validate-update-validate-update-validate chain.